Based on the commit below, with minor adjustments.
https://github.com/JetBrains/kotlin/commit/9cb6ad8437f77a7f6a51a877c327dbb691c345ae
http://youtrack.jetbrains.com/issue/KT-81108
Bug: 444514435

[LL] Introduce custom LLFirModuleData for native forward declarations
- The previously used `FirBinaryDependenciesModuleData` is not an
  `LLFirModuleData`, which leads to a class cast exception when trying
  to get the module data of a forward declaration FIR element in lazy
  resolution. All FIR elements in Analysis API mode must have
  `LLFirModuleData`.
- We can instantiate a custom `LLFirModuleData` which keeps the
  `FORWARD_DECLARATIONS_MODULE_NAME`. The module data has the same
  `KaModule` as the session since the forward declarations are
  conceptually still part of the same module.
- It's currently difficult to add a test for this as we lack the
  infrastructure for "true" native forward declarations, but I tested
  the fix locally with Dokka.

^KT-81108 fixed

diff --git a/analysis/low-level-api-fir/src/org/jetbrains/kotlin/analysis/low/level/api/fir/projectStructure/LLFirModuleData.kt b/analysis/low-level-api-fir/src/org/jetbrains/kotlin/analysis/low/level/api/fir/projectStructure/LLFirModuleData.kt
index 66497b99ef7..7ca03192760 100644
--- a/analysis/low-level-api-fir/src/org/jetbrains/kotlin/analysis/low/level/api/fir/projectStructure/LLFirModuleData.kt
+++ b/analysis/low-level-api-fir/src/org/jetbrains/kotlin/analysis/low/level/api/fir/projectStructure/LLFirModuleData.kt
@@ -30,7 +30,10 @@ val FirBasedSymbol<*>.llFirModuleData: LLFirModuleData
     get() = fir.llFirModuleData
 
 
-class LLFirModuleData private constructor(val ktModule: KaModule) : FirModuleData() {
+/**
+ * The [FirModuleData] for FIR elements managed by the Analysis API. In Analysis API mode, all FIR elements must have [LLFirModuleData].
+ */
+open class LLFirModuleData internal constructor(val ktModule: KaModule) : FirModuleData() {
     constructor(session: LLFirSession) : this(session.ktModule) {
         bindSession(session)
     }
diff --git a/analysis/low-level-api-fir/src/org/jetbrains/kotlin/analysis/low/level/api/fir/symbolProviders/factories/LLBinaryOriginLibrarySymbolProviderFactory.kt b/analysis/low-level-api-fir/src/org/jetbrains/kotlin/analysis/low/level/api/fir/symbolProviders/factories/LLBinaryOriginLibrarySymbolProviderFactory.kt
index 57a5a937f52..13e7d866f78 100644
--- a/analysis/low-level-api-fir/src/org/jetbrains/kotlin/analysis/low/level/api/fir/symbolProviders/factories/LLBinaryOriginLibrarySymbolProviderFactory.kt
+++ b/analysis/low-level-api-fir/src/org/jetbrains/kotlin/analysis/low/level/api/fir/symbolProviders/factories/LLBinaryOriginLibrarySymbolProviderFactory.kt
@@ -8,6 +8,7 @@ package org.jetbrains.kotlin.analysis.low.level.api.fir.symbolProviders.factorie
 import com.intellij.openapi.diagnostic.Logger
 import com.intellij.psi.search.GlobalSearchScope
 import org.jetbrains.kotlin.analysis.api.projectStructure.KaLibraryModule
+import org.jetbrains.kotlin.analysis.api.projectStructure.KaModule
 import org.jetbrains.kotlin.analysis.low.level.api.fir.projectStructure.LLFirModuleData
 import org.jetbrains.kotlin.analysis.low.level.api.fir.projectStructure.moduleData
 import org.jetbrains.kotlin.analysis.low.level.api.fir.sessions.LLFirSession
@@ -31,6 +32,7 @@ import org.jetbrains.kotlin.library.metadata.impl.KlibResolvedModuleDescriptorsF
 import org.jetbrains.kotlin.load.kotlin.PackageAndMetadataPartProvider
 import org.jetbrains.kotlin.load.kotlin.PackagePartProvider
 import org.jetbrains.kotlin.load.kotlin.VirtualFileFinderFactory
+import org.jetbrains.kotlin.name.Name
 import org.jetbrains.kotlin.util.Logger as KLogger
 import org.jetbrains.kotlin.utils.exceptions.rethrowIntellijPlatformExceptionIfNeeded
 import java.nio.file.Path
@@ -92,7 +94,7 @@ internal object LLBinaryOriginLibrarySymbolProviderFactory : LLLibrarySymbolProv
     ): List<FirSymbolProvider> {
         val moduleData = session.moduleData
         val moduleDataProvider = SingleModuleDataProvider(moduleData)
-        val forwardDeclarationsModuleData = FirBinaryDependenciesModuleData(FORWARD_DECLARATIONS_MODULE_NAME).apply {
+        val forwardDeclarationsModuleData = LLNativeForwardDeclarationsModuleData(moduleData.ktModule).apply {
             bindSession(session)
         }
 
@@ -104,6 +106,16 @@ internal object LLBinaryOriginLibrarySymbolProviderFactory : LLLibrarySymbolProv
         )
     }
 
+    /**
+     * The module data specifically for [originalModule]'s associated native forward declarations. In particular, this module data is an
+     * [LLFirModuleData]-compliant replacement for the [FirBinaryDependenciesModuleData] used on the compiler side (see
+     * `FirNativeSessionFactory.createAdditionalDependencyProviders`).
+     */
+    private class LLNativeForwardDeclarationsModuleData(originalModule: KaModule) : LLFirModuleData(originalModule) {
+        override val name: Name
+            get() = FORWARD_DECLARATIONS_MODULE_NAME
+    }
+
     override fun createJsLibrarySymbolProvider(
         session: LLFirSession,
         scope: GlobalSearchScope,
